src/ai/backend/manager/repositories/permission_controller/db_source.py

```diff
@@ -142,3 +146,271 @@ async def get_entity_mapped_scopes(
             )
             result = await db_session.scalars(stmt)
             return result.all()
+
+    async def get_scope_permissions(
+        self,
+        user_id: uuid.UUID,
+        scope_id: ScopeId,
+    ) -> ScopePermissionSet:
+        role_query = (
+            sa.select(RoleRow)
+            .select_from(
+                sa.join(UserRoleRow, RoleRow.id == UserRoleRow.role_id).join(
+                    PermissionGroupRow, RoleRow.id == PermissionGroupRow.role_id
+                )
+            )
+            .where(
+                sa.and_(
+                    RoleRow.status == RoleStatus.ACTIVE,
+                    UserRoleRow.user_id == user_id,
+                    sa.or_(
+                        PermissionGroupRow.scope_type == ScopeType.GLOBAL,
+                        PermissionGroupRow.scope_id == scope_id.scope_id,
+                    ),
+                )
+            )
+            .options(
+                contains_eager(RoleRow.permission_group_rows).options(
+                    selectinload(PermissionGroupRow.permission_rows)
+                )
+            )
+        )
+        async with self._db.begin_readonly_session() as db_session:
+            role_rows = await db_session.scalars(role_query)
+            role_rows = cast(list[RoleRow], role_rows.all())
+
+            scope_permissions: set[OperationType] = set()
+            global_permissions: Optional[set[OperationType]] = None
+            for role in role_rows:
+                for pg in role.permission_group_rows:
+                    if pg.parsed_scope_id() != scope_id:
+                        continue
+                    if pg.scope_type == ScopeType.GLOBAL:
+                        global_permissions = {perm.operation for perm in pg.permission_rows}
+                    else:
+                        scope_permissions |= {perm.operation for perm in pg.permission_rows}
+
+        return ScopePermissionSet(scope_id, scope_permissions, global_permissions)
+
+    async def check_scope_permission_exist(
+        self,
+        user_id: uuid.UUID,
+        scope_id: ScopeId,
+        operation: OperationType,
+    ) -> bool:
```

Comment:
If we are returning the permission set anyway, it doesn't seem necessary to have a separate check function.